cachito/workers/pkg_managers/yarn.py

```diff
@@ -55,3 +64,293 @@ def get_yarn_proxy_repo_username(request_id):
+        yarn_lock_replaced = _replace_deps_in_yarn_lock(yarn_lock, nexus_replacements)
+    else:
+        package_json_replaced = None
+        yarn_lock_replaced = None
+
+    return {
+        "package": package,
+        "deps": deps,
+        "package.json": package_json_replaced,
+        "lock_file": yarn_lock_replaced,
+    }
+
+
+def _replace_deps_in_package_json(package_json, nexus_replacements):
+    """
+    Replace non-registry dependencies in package.json with their versions in Nexus.
+
+    :param dict package_json: parsed package.json data
+    :param dict nexus_replacements: modified subset of yarn.lock data, a dict in the format:
+        {<dependency identifier>: <dependency info>}
+    :return: copy of package.json data with replacements applied (or None if no replacements match)
+    """
+    expanded_replacements = {}
+
+    for deps_identifier, nexus_replacement in nexus_replacements.items():
+        # The dependency identifier is a line that may contain multiple comma-separated
+        # dependencies (different ways to specify the same dependency)
+        deps = map(str.strip, deps_identifier.split(","))
+        for d in deps:
+            expanded_replacements[d] = nexus_replacement
+
+    package_json_new = copy.deepcopy(package_json)
+    modified = False
+
+    # TODO: optionalDependencies also exist, ignored for npm, safe to ignore for yarn?
+    for dep_type in ("dependencies", "devDependencies"):
+        for dep_name, dep_version in package_json.get(dep_type, {}).items():
+            dep_identifier = f"{dep_name}@{dep_version}"
+
+            if dep_identifier not in expanded_replacements:
+                continue
+
+            new_version = expanded_replacements[dep_identifier]["version"]
+            log.info(
+                "Replacing the version of %s in %s from %s to %s in package.json",
+                dep_name,
+                dep_type,
+                dep_version,
+                new_version,
+            )
+            package_json_new[dep_type][dep_name] = new_version
+            modified = True
+
+    return package_json_new if modified else None
+
+
+def _replace_deps_in_yarn_lock(yarn_lock, nexus_replacements):
+    """
+    Replace non-registry dependencies in yarn.lock with their versions in Nexus.
+
+    :param dict yarn_lock: parsed yarn.lock data
+    :param dict nexus_replacements: modified subset of yarn.lock data, a dict in the format:
+        {<dependency identifier>: <dependency info>}
+    :return: copy of yarn.lock data with replacements applied
+    """
+    yarn_lock_new = copy.deepcopy(yarn_lock)
+    yarn_lock_new.update(copy.deepcopy(nexus_replacements))
+    return yarn_lock_new
+
+
+def _sanity_check(app_source_path):
+    """
+    Sanity-check the application directory.
+
+    - package.json must be present
+    - yarn.lock must be present
+    - package-lock.json must not be present
+    # TODO: npm-shrinkwrap.json?
```

Comment:
It seems this one should not be present either.

Reply:
Oh, I was actually going to remove this function altogether in favour of [doing this check in the fetch_yarn_source task](https://github.com/chmeliik/cachito/blob/522560f7abf70ab4c5addd6023775c93d133e162/cachito/workers/tasks/yarn.py#L63). Forgot to do that, my bad (but this answers the question that I would have had in the future anyway :D )